Add additional track pieces to LIM Launched Coaster and the Looping Coaster

X123M3-256 opened the pull request:
This PR adds additional track pieces to the LIM/Looping coaster track. The following track pieces are made available by default on the LIM launched coaster:

* Large flat to steep
* Level banked turn to unbanked slope
* Small half loops
* Medium half loops
* Large half loops
* Large corkscrews
* Barrel rolls

The following track pieces are added to the looping coaster:

* Large steep to flat
* Level banked turn to unbanked slope

The following track pieces are also added, but not made available by default:

* Small zero G roll
* Large zero G roll
* Small flat to steep

karst commented:
ImageId has recently been refactored in #18126, might be that the build is failing because of that.

Generalcamo commented on src/openrct2/ride/coaster/meta/LIMLaunchedRollerCoaster.h:
```diff
@@ -19,8 +19,8 @@ constexpr const RideTypeDescriptor LIMLaunchedRollerCoasterRTD =
     SET_FIELD(Category, RIDE_CATEGORY_ROLLERCOASTER),
-    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_FLAT_ROLL_BANKING, TRACK_VERTICAL_LOOP, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_TWIST, TRACK_CORKSCREW, TRACK_HELIX_SMALL, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_SLOPE_VERTICAL, TRACK_BLOCK_BRAKES, TRACK_SL ...
-    SET_FIELD(ExtraTrackPieces, {}),
+    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_FLAT_ROLL_BANKING, TRACK_VERTICAL_LOOP, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_TWIST, TRACK_CORKSCREW, TRACK_HELIX_SMALL, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_SLOPE_VERTICAL, TRACK_BLOCK_BRAKES, TRACK_SL ...
```
The addition of the barrel roll to the LIM coaster should be noted in the changelog

Gymnasiast replied:
Please leave these things to us.

Generalcamo commented on src/openrct2/ride/coaster/meta/LoopingRollerCoaster.h:
```diff
@@ -19,8 +19,8 @@ constexpr const RideTypeDescriptor LoopingRollerCoasterRTD =
-    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_LIFT_HILL, TRACK_FLAT_ROLL_BANKING, TRACK_VERTICAL_LOOP, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_HELIX_SMALL, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_BLOCK_BRAKES, TRACK_SLOPE_ROLL_BANKING, TRACK_BOOSTER}),
-    SET_FIELD(ExtraTrackPieces, {TRACK_SLOPE_VERTICAL, TRACK_TWIST, TRACK_CORKSCREW, TRACK_CURVE_VERTICAL, TRACK_QUARTER_LOOP}),
+    SET_FIELD(EnabledTrackPieces, {TRACK_STRAIGHT, TRACK_STATION_END, TRACK_LIFT_HILL, TRACK_FLAT_ROLL_BANKING, TRACK_VERTICAL_LOOP, TRACK_SLOPE, TRACK_SLOPE_STEEP_UP, TRACK_SLOPE_STEEP_DOWN, TRACK_SLOPE_CURVE, TRACK_SLOPE_CURVE_STEEP, TRACK_S_BEND, TRACK_CURVE_SMALL, TRACK_CURVE, TRACK_HELIX_SMALL, TRACK_BRAKES, TRACK_ON_RIDE_PHOTO, TRACK_BLOCK_BRAKES, TRACK_SLOPE_ROLL_BANKING, TRACK_BOOSTER, TR ...
+    SET_FIELD(ExtraTrackPieces, {TRACK_SLOPE_VERTICAL, TRACK_TWIST, TRACK_CORKSCREW, TRACK_CURVE_VERTICAL, TRACK_QUARTER_LOOP, TRACK_SLOPE_LONG, TRACK_HALF_LOOP, TRACK_HALF_LOOP_MEDIUM, TRACK_HALF_LOOP_LARGE,TRACK_CORKSCREW_LARGE}),
```
If we're including the large corkscrew into the extra track pieces, and we already have the twist available here, should probably include the barrel roll too.

X123M3-256 replied:
Yeah, this was a mistake - the zero G rolls were also missing. It has been fixed now.

Generalcamo reviewed (commented):
Some sprites for booster pieces are included for the LIM, but I notice no changes from the default  straight track pieces. Was this an accidental inclusion?

X123M3-256 commented:
> Some sprites for booster pieces are included for the LIM, but I notice no changes from the default straight track pieces. Was this an accidental inclusion?

Yes, the original plan was to add the booster sprites as part of this PR so I added some placeholder sprites. I've now decided to do that in a separate PR as I don't know how long it'll take to come up with a suitable booster sprite, and I'm thinking of adding vertical boosters as a new track piece.

I've now removed the unused booster sprites and associated sprite IDs from this branch.

Gymnasiast commented:
Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?

X123M3-256 commented:
> Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?

Is there any reason not to? It doesn't require any sprites that the looping trains don't have and it's a useful track piece to have.

Gymnasiast commented:
> > Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?
> 
> Is there any reason not to? It doesn't require any sprites that the looping trains don't have and it's a useful track piece to have.

To me, it doesn’t seem to fit the rest of the coaster type, considering it’s usually compact, twisty designs.

spacek531 commented:
> > > Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?
> > 
> > 
> > Is there any reason not to? It doesn't require any sprites that the looping trains don't have and it's a useful track piece to have.
> 
> To me, it doesn’t seem to fit the rest of the coaster type, considering it’s usually compact, twisty designs.

This decision should be made by the user, in my opinion.

Gymnasiast commented:
> > > > Did you have a specific reason to enable ”Large steep to flat” pieces on the Looping RC by default?
> > > 
> > > 
> > > Is there any reason not to? It doesn't require any sprites that the looping trains don't have and it's a useful track piece to have.
> > 
> > 
> > To me, it doesn’t seem to fit the rest of the coaster type, considering it’s usually compact, twisty designs.
> 
> This decision should be made by the user, in my opinion.

We have hidden more track elements behind this flag that vehicles are technically capable of, but that would be unrealistic. This includes stuff like the holding brake on the Twister RC.

IMO, the user already has a way to decide and that is to turn on the cheat.

Incidentally, this is a minor thing and we can very easily revisit the decision at any time.

X123M3-256 commented:
> We have hidden more track elements behind this flag that vehicles are technically capable of, but that would be unrealistic. 

This isn't a piece that would be unrealistic. It's not really unrealistic for any ride type except perhaps the wild mouse (and of course those that don't have steep slopes at all) - the standard size slope transitions are very small. 

Arguably, the medium and large loops are realistic as well, but I left those disabled as I don't think they really fit.

ZeeMajora commented:
I think the large steep to flat pieces are fine on the looping RC and don't really look that out of place to have them available by default, you can already kind of fake them with the normal incline pieces this just looks much nicer.

![Screenshot_select-area_20221002181251](https://user-images.githubusercontent.com/42477864/193478591-3364633a-bc3d-40fd-9c0d-125d982f8622.png)

Gymnasiast commented:
Fine, let’s leave it enabled by default then.

Gymnasiast commented:
I did a quick test and it seems to work nicely.  I have asked @rik-smeets to test it as well and make sure there is no z-fighting, clipping or anything like that.

rik-smeets reviewed (changes requested):
I noticed some issues, could you try to fix these? You can use this park to reproduce:
[LIM new track pieces.zip](https://github.com/OpenRCT2/OpenRCT2/files/9708469/LIM.new.track.pieces.zip)

1. Clipping on large flat to steep
![Clipping large flat to steep](https://user-images.githubusercontent.com/30838294/193863534-904f32a8-cff3-4f4c-8009-b3ec66397fd5.gif)

2. Zero G rolls
Trains passing through the zero G rolls don't look exactly right, clipping through the tracks. In case of the large zero G roll, also some clipping through the tree.
![Large Zero G](https://user-images.githubusercontent.com/30838294/193863717-84dc9f58-a95b-440c-b23a-203001c2a159.gif)
![Small Zero G](https://user-images.githubusercontent.com/30838294/193863746-5506335f-da6b-46e6-b3a0-08be5a86c3e2.gif)

3. Disconnected barrel roll on flat land
![image](https://user-images.githubusercontent.com/30838294/193863910-bdc0f5ec-63a0-4bbf-b4d4-42f82f81eadd.png)

karst commented:
> Zero G rolls
Trains passing through the zero G rolls don't look exactly right, clipping through the tracks. In case of the large zero G roll, also some clipping through the tree.

Could you test this with a train that supports zero-g rolls?

karst commented:
![image](https://user-images.githubusercontent.com/1516893/193869329-fb90d71e-c20c-4e62-994a-256f91f948e0.png)


The first one is also a train related bug, I imagine that'd have to be fixed in the objects directory.

X123M3-256 commented:
@rik-smeets 

> Clipping on large flat to steep

I can't figure out the cause of this issue, but I don't think it's related to the track piece -  if you put the LIM trains on giga or twister track, they still clip through the tree. In fact, they still clip with the track made invisible. But if you put the giga or twister trains on the LIM track, they don't clip. The bounding boxes for this track piece were copied from the giga coaster so I wouldn't expect different behaviour.

The fact that the other train types don't have a problem suggests it's an issue with the vehicle, but unlike other vehicle related glitches, the bounding boxes look reasonable so I'm not sure why this happens. In any case, I don't think any changes to the track piece will prevent it.

> Trains passing through the zero G rolls don't look exactly right, clipping through the tracks. In case of the large zero G roll, also some clipping through the tree.

The clipping through the tree is the same issue as on the large steep to flat. I'm not sure what you mean about the trains ... of course they're not going to look right when you use a train that can't do zero G rolls, did you test with the hybrid or single rail train?

> Disconnected barrel roll on flat land

That one definitely is a problem with the track piece, I'll push a fix in a minute.

Gymnasiast commented:
@rik-smeets Could you repeat the clipping test using the Looping RC train?

X123M3-256 commented:
The barrel roll issue should be fixed now

rik-smeets commented:
@Gymnasiast The Looping RC trains don't clip, so as @karst said it seems to be an issue with the vehicles. I'll create a separate issue for it.

@X123M3-256 You're right, as @Gymnasiast and @karst said: I was mistakenly using the wrong train type. Sorry for the confusion.

I'll retest the barrel roll with the Windows build as soon as it's completed, thanks for the quick action!

rik-smeets commented:
The issue is fixed, but the barrel roll (left) doesn't properly connect to the adjacent piece of track:
![image](https://user-images.githubusercontent.com/30838294/193891798-4017523b-dc00-4fba-b912-49ef7053b3b1.png)
![image](https://user-images.githubusercontent.com/30838294/193892247-be56fc9c-106f-4dce-aee6-ed5398f3de1d.png)


This doesn't happen with the barrel roll (right). Could you fix this too @X123M3-256 ?

SpartanFrederic104 commented:
Does it happen only on this specific angle?

rik-smeets commented:
> Does it happen only on this specific angle?

Yes, I only noticed it at that angle.

X123M3-256 commented:
> This doesn't happen with the barrel roll (right). Could you fix this too X123M3-256 ?

I already went through every sprite and edited them by hand to try and smooth out the transitions as much as I could but I'm wondering if I missed that one, because it does look like it could be made a little smoother than it is. 

I'll take another look at it, but it's impossible to make every piece line up because the existing pieces don't all match up with each other. For example, the flat to gentle slope pieces don't line up with the straight gentle slope pieces, so the half loops won't always line up if you place a gentle slope piece before (I decided that the flat to gentle into a half loop would be the more common situation). Because of this, there's quite a few combinations of vanilla pieces that don't connect smoothly too.

X123M3-256 commented:
OK, I've touched up that sprite a bit, that's probably the best I can do with it. It looks better than before, I think.

rik-smeets commented:
> OK, I've touched up that sprite a bit, that's probably the best I can do with it. It looks better than before, I think.

It does look better, thanks!
